skcosmo/pcovr/kpcovr.py

```diff
@@ -0,0 +1,383 @@
+        return self.kernel == "precomputed"
+
+    def _get_kernel(self, X, Y=None):
+        if callable(self.kernel):
+            params = self.kernel_params or {}
+        else:
+            params = {"gamma": self.gamma, "degree": self.degree, "coef0": self.coef0}
+        return pairwise_kernels(
+            X, Y, metric=self.kernel, filter_params=True, n_jobs=self.n_jobs, **params
+        )
+
+    def _fit(self, K, Yhat, W):
+        """
+        Fit the model with the computed kernel and approximated properties.
+        """
+
+        K = self._centerer.fit_transform(K)
+        K_tilde = pcovr_kernel(mixing=self.mixing_, X=K, Y=Yhat, kernel="precomputed")
+
+        v, U = eig_solver(
+            K_tilde, tol=self.regularization_, n_components=self.n_components
+        )
+
+        P = (self.mixing_ * np.eye(K.shape[0])) + (1.0 - self.mixing_) * (W @ Yhat.T)
+
+        v_inv = np.linalg.pinv(np.diagflat(v))
+
+        self.pkt_ = P @ U @ np.sqrt(v_inv)
+
+        T = K @ self.pkt_
+
+        self.pt__ = np.linalg.pinv(T.T @ T) @ T.T
+
+    def fit(self, X, Y, Yhat=None, W=None):
+        """
+
+        Fit the model with X and Y.
+
+        Parameters
+        ----------
+        X: array-like, shape (n_samples, n_features)
+            Training data, where n_samples is the number of samples and
+            n_features is the number of features.
+
+            It is suggested that :math:`\\mathbf{X}` be centered by its column-
+            means and scaled. If features are related, the matrix should be scaled
+            to have unit variance, otherwise :math:`\\mathbf{X}` should be
+            scaled so that each feature has a variance of 1 / n_features.
+
+        Y: array-like, shape (n_samples, n_properties)
+            Training data, where n_samples is the number of samples and
+            n_properties is the number of properties
+
+            It is suggested that :math:`\\mathbf{X}` be centered by its column-
+            means and scaled. If features are related, the matrix should be scaled
+            to have unit variance, otherwise :math:`\\mathbf{Y}` should be
+            scaled so that each feature has a variance of 1 / n_features.
+
+        Yhat: array-like, shape (n_samples, n_properties), optional
+            Regressed training data, where n_samples is the number of samples and
+            n_properties is the number of properties. If not supplied, computed
+            by ridge regression.
+
+        Returns
+        -------
+        self: object
+            Returns the instance itself.
+
+        """
+
+        X, Y = check_X_y(X, Y, y_numeric=True, multi_output=True)
+        self.X_fit_ = X.copy()
+
+        if self.n_components is None:
+            self.n_components = min(X.shape)
+
+        self._centerer = KernelFlexibleCenterer()
+        K = self._get_kernel(X)
+
+        if W is None:
+            if Yhat is None:
+                W = (np.linalg.pinv(K, rcond=self.regularization_) @ Y).reshape(
```

Comment:
```suggestion
                W = (np.linalg.lstsq(K, Y, rcond=self.regularization_)[0]).reshape(
```
This is probably a good spot for a `np.linalg.lstsq` regardless of what we decide to do about the `p` matrices. On a similar note, is there a reason why this is a `pinv` and the one in the `else` below is a `lstsq`?

Reply:
Only reason is me breaking out of the habit of pinv...